src/EditorFeatures/Core/DefinitionContextTracker.cs

```diff
@@ -0,0 +1,211 @@
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Editor;
+using Microsoft.VisualStudio.Utilities;
+
+namespace Microsoft.CodeAnalysis.Editor
+{
+    [Export(typeof(IWpfTextViewConnectionListener))]
+    [ContentType(ContentTypeNames.RoslynContentType)]
+    [TextViewRole(PredefinedTextViewRoles.Interactive)]
+    internal class DefinitionContextConnectionListener : IWpfTextViewConnectionListener
+    {
+        private readonly Dictionary<ITextView, CancellationTokenSource> _views = new Dictionary<ITextView, CancellationTokenSource>();
+        private readonly IMetadataAsSourceFileService _metadataAsSourceFileService;
+        private readonly ICodeDefinitionWindowService _codeDefinitionWindowService;
+
+        [ImportingConstructor]
+        public DefinitionContextConnectionListener(
+            IMetadataAsSourceFileService metadataAsSourceFileService,
+            ICodeDefinitionWindowService codeDefinitionWindowService)
+        {
+            _metadataAsSourceFileService = metadataAsSourceFileService;
+            _codeDefinitionWindowService = codeDefinitionWindowService;
+        }
+
+        void IWpfTextViewConnectionListener.SubjectBuffersConnected(IWpfTextView textView, ConnectionReason reason, Collection<ITextBuffer> subjectBuffers)
+        {
+            if (!_views.ContainsKey(textView) && !textView.Roles.Contains(PredefinedTextViewRoles.CodeDefinitionView))
+            {
+                _views.Add(textView, new CancellationTokenSource());
+                textView.Caret.PositionChanged += OnTextViewCaretPositionChanged;
+                var fireAndForget = UpdateDefinitionContext(textView, textView.Caret.Position);
+            }
+        }
+
+        void IWpfTextViewConnectionListener.SubjectBuffersDisconnected(IWpfTextView textView, ConnectionReason reason, Collection<ITextBuffer> subjectBuffers)
+        {
+            if (reason == ConnectionReason.TextViewLifetime ||
+                !textView.BufferGraph.GetTextBuffers(b => b.ContentType.IsOfType(ContentTypeNames.RoslynContentType)).Any())
+            {
+                CancellationTokenSource cancellationTokenSource;
+                if (_views.TryGetValue(textView, out cancellationTokenSource))
+                {
+                    cancellationTokenSource.Cancel();
+                    _views.Remove(textView);
+                    textView.Caret.PositionChanged -= OnTextViewCaretPositionChanged;
+                }
+            }
+        }
+
+        private void OnTextViewCaretPositionChanged(object sender, CaretPositionChangedEventArgs e)
+        {
+            var fireAndForget = UpdateDefinitionContext(e.TextView, e.NewPosition);
+        }
+
+        private async Task UpdateDefinitionContext(ITextView textView, CaretPosition caretPosition)
+        {
+            // Cancel any pending update for this view
+            _views[textView].Cancel();
```

Comment:
you need to check for cancellations inside `GetContextFromPoint` for this to be effective?